htmd/smallmol/chemlab/builder.py

```diff
@@ -0,0 +1,406 @@
+
+            nbr1Vect = _normalizeVector(nbr1Vect)
+            nbr2Vect = _normalizeVector(nbr2Vect)
+
+            dirVect = nbr1Vect + nbr2Vect
+            dirVect = _normalizeVector(dirVect)
+
+            if heavy_hybridization == 4:
+                nbrPerp = np.cross(nbr1Vect, nbr2Vect)
+                rotAxis = np.cross(nbrPerp, dirVect)
+                rotAxis = _normalizeVector(rotAxis)
+
+                rot_matrix = _getRotationMatrix(rotAxis, (109.471/2), deg=True)
+
+                dirVect = np.dot(rot_matrix, dirVect)
+
+        elif heavy_neighbors_num == 4:
+          #  print('last attachment sp3 procedure')
+            nbrs = self._filterNeighbors(attachToIdx, atomIdx)
+            nbr1, nbr2, nbr3 = nbrs
+
+            nbr1_coords, nbr2_coords, nbr3_coords = sm.coords[nbr1].reshape(3), sm.coords[nbr2].reshape(3), sm.coords[nbr3].reshape(3)
+
+            nbr1Vect = heavy_coords - nbr1_coords
+            nbr2Vect = heavy_coords - nbr2_coords
+            nbr3Vect = heavy_coords - nbr3_coords
+
+            nbr1Vect = _normalizeVector(nbr1Vect)
+            nbr2Vect = _normalizeVector(nbr2Vect)
+            nbr3Vect = _normalizeVector(nbr3Vect)
+
+            dirVect = nbr1Vect + nbr2Vect + nbr3Vect
+            dirVect = _normalizeVector(dirVect)
+
+        atom_coords = heavy_coords + dirVect * bondLength
+
+        return  atom_coords.reshape(3,1)
+
+    def _filterNeighbors(self, atomIdx, exclude):
+        """
+        Returns all the indexes of the neighbors atom except for the excluded one
+
+        Parameters
+        ----------
+        atomIdx: int
+         The index of the central atom
+        exclude: int
+            The index of the atom to exclude
+
+        Returns
+        --------
+        neighbors: list
+            A list of the neighbors
+
+        """
+
+        sm = self.smallmol
+
+        neighbors = sm.neighbors[atomIdx]
+        return [ n for n in neighbors if n != exclude]
+
+    def _removeAtoms(self, ids):
+        """
+        Remove the atoms based on the indexes provided
+
+        Paraemters
+        ----------
+        ids: list
+            A list of atom indexes to remove
+        """
+
+        sm = self.smallmol
+
+        for k in sm._atom_fields:
+
+            if k == 'coords':
+                sm.__dict__[k] = np.delete(sm.__dict__[k], ids, axis=0)
+            else:
+                sm.__dict__[k] = np.delete(sm.__dict__[k], ids)
+
+        new_neighbors = []
+        new_bondstype = []
+        for n_set, b_set in zip(sm.__dict__['neighbors'], sm.__dict__['bondtypes']):
+            new_n_set = []
+            new_b_set = []
+            for n, b in zip(n_set, b_set):
+                if n not in ids:
+                    new_n_set.append(n)
+                    new_b_set.append(b)
+            new_neighbors.append(new_n_set)
+            new_bondstype.append(new_b_set)
+        sm.__dict__['neighbors'] = np.array(new_neighbors)
```

Comment:
No need for `__dict__` anywhere in this function except in the first loop.

Reply:
done it.